modules/imgproc/src/imgwarp.cpp

```diff
@@ -2307,6 +2307,7 @@ class WarpAffineInvoker :
                             opt_SSE4_1::WarpAffineInvoker_Blockline_SSE41(adelta + x, bdelta + x, xy, X0, Y0, bw);
                         else
                         #endif
+                        if( cv_hal_warpAffineBlocklineNN(adelta + x, bdelta + x, xy, X0, Y0, bw) != CV_HAL_ERROR_OK )
```

Comment:
Please extract the whole block to some `cv::hal::` function. Define it here: https://github.com/opencv/opencv/blob/4.x/modules/imgproc/include/opencv2/imgproc/hal/hal.hpp Implementation can reside somewhere in this file (imgwarp.cpp).

This new function should first try external HAL function (using [CALL_HAL](https://github.com/opencv/opencv/blob/8d935e21843c1085000ecaf1ec46337c7891f651/modules/core/src/hal_replacement.hpp#L971-L990) macro), then try AVX, SSE, LASX, universal intrinsics, then fallback implementation.

Then `cv::warpAffine` should call this new `cv::hal::` function for CPU processing.

-----------

There are some functions implemented this way, e.g. `cv::hal::normHamming` (https://github.com/opencv/opencv/blob/8d935e21843c1085000ecaf1ec46337c7891f651/modules/core/src/norm.cpp#L53-L100), but it only have HAL, universal intrinsics and fallback parts.

Reply:
4 new functions have been extracted.  Further accuracy checks might be needed for related platforms.